src/Integration.Vsix/SonarLintTagger/AccumulatingIssueConsumer.cs

```diff
@@ -0,0 +1,117 @@
+ * 1) we're triggering analysis on save, so the state of the text buffer matches the file on disc, and
+ * 2) the file won't have had time to change again before the analyzer reads it so we expect always to
+ *    be able to map back from the reported issue to the snapshot.
+ * However, we code defensively and strip out any issues that can't be mapped to the analysis snapshot.
+ * 
+ */
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    /// <summary>
+    /// Handles processing the issues for a single analysis request
+    /// </summary>
+    internal class AccumulatingIssueConsumer : IIssueConsumer
+    {
+        // See bug #1487: this text snapshot should match the content of the file being analysed
+        private readonly ITextSnapshot analysisSnapshot;
+        private readonly IIssueToIssueMarkerConverter issueMarkerConverter;
+        private readonly List<IssueMarker> allIssueMarkers;
+        private readonly string analysisFilePath;
+        private readonly OnIssuesChanged onIssuesChanged;
+
+        public delegate void OnIssuesChanged(IEnumerable<IssueMarker> issueMarkers);
+
+        public AccumulatingIssueConsumer(ITextSnapshot analysisSnapshot, string analysisFilePath, OnIssuesChanged onIssuesChangedCallback)
+            : this(analysisSnapshot, analysisFilePath, onIssuesChangedCallback, new IssueToIssueMarkerConverter())
+        {
+        }
+
+        internal /* for testing */ AccumulatingIssueConsumer(ITextSnapshot analysisSnapshot, string analysisFilePath, OnIssuesChanged onIssuesChangedCallback,
+            IIssueToIssueMarkerConverter issueMarkerConverter)
+        {
+            this.analysisSnapshot = analysisSnapshot ?? throw new ArgumentNullException(nameof(analysisSnapshot));
+            this.analysisFilePath = analysisFilePath ?? throw new ArgumentNullException(nameof(analysisFilePath));
+            this.onIssuesChanged = onIssuesChangedCallback ?? throw new ArgumentNullException(nameof(onIssuesChangedCallback));
+
+            this.issueMarkerConverter = issueMarkerConverter ?? throw new ArgumentNullException(nameof(analysisSnapshot));
+
+            allIssueMarkers = new List<IssueMarker>();
+        }
+
+        public void Accept(string path, IEnumerable<IAnalysisIssue> issues)
+        {
+            // Callback from the daemon when new results are available
+            if (path != analysisFilePath)
+            {
+                Debug.Fail("Issues returned for an unexpected file path");
+                return;
+            }
+
+            Debug.Assert(issues.All(IsIssueInAnalysisSnapshot), "Not all reported issues could be mapped to the analysis snapshot");
+
+            var newMarkers = issues
+                .Where(IsIssueInAnalysisSnapshot)
+                .Select(x => issueMarkerConverter.Convert(x, analysisSnapshot))
+                .ToArray();
+
+            allIssueMarkers.AddRange(newMarkers);
+
+            onIssuesChanged.Invoke(allIssueMarkers);
```

Comment:
There is a problem here: we invoke this callback time a new issue arrives, and we hand it all accumulated issues so far. Which means that we keep repeating the same logic over and over again: checking for suppressions and calculating spans for issues that have already been calculated. As long as the text buffer has not been changed, there is no reason to keep re-calculating this data.
This logic should move out of the TextBufferIssueTracker as well:
1. IIssuesFilter should be called from AccumulatingIssueConsumer
2. TextBufferIssueTracker.TranslateSpans method should be extracted to a service, and called from both AccumulatingIssueConsumer (when issues arrive) and from TextBufferIssueTracker (when the buffer changes). It would also simplify the tests this way.

Reply:
1. I deliberately didn't do the filtering here precisely so the original list of issues was preserved. The issue filtering should be "live" i.e. if we have fetched new suppressions from the server in the meantime then what is filtered out could change. The TextBufferIssueTracker is also filtering too early, but that's a separate issue.

2. This class only maps the "raw" issues to spans in the original analysis snapshot, and it only maps spans for the newly-received issues. The issues have to be mapped against the analysis snapshot (to fix #1487).
This class has no concept of the notion of _translating_ spans between different snapshots. This is deliberate - only the text buffer needs to know that multiple snapshots exist, and which one is the current snapshot.

The list of accumulated markers is then sent to the TextBufferIssueTracker, which may or may not have a changed in the meantime. If it has changed, then all of the spans need to be _translated_.

This implementation is functionally correct, and I chose it as it's the simplest one that could work. I agree that at present it may be translating spans repeatedly in some cases. However, translating spans has to be very fast, as it happens every time to press a key in the editor. Since it doesn't seem to be causing a performance problem, do we need to do anything more complicated?

Reply:
To the outside reviewer, I disagree that this is the simplest implementation. 
The snapshot logic is complicated, and it being split between the two classes (AccumulatingIssueConsumer and TextBufferIssueTracker). 
The fact that AccumulatingIssueConsumer calculates the spans of the issues for _a snapshot_ and not _the current snapshot_ is extremely confusing. It's doing half a job. The calculation is incomplete.

Either AccumulatingIssueConsumer should know about the translation, or the TextBufferIssueTracker should only know about markers and not issues -- which means that the suppression logic shouldn't be there.
If we say that TextBufferIssueTracker is responsible for translating markers to the latest snapshot, then that's all it should do.

Reply:
As discussed offline (in case we ever need to dig out this PR again in the future)
* the issue consumer knows about how to map from file-level issues to the in-memory snapshot representation. It doesn't need to know that multiple snapshots exist.
* the text buffer already knows that multiple snapshots exists, and mapping between them is part of its job.
So I think the division of responsibilities with regard to the spans makes sense.

Where the filtering is done is an orthogonal issue. It could be done in either, and at this point it probably doesn't matter much which.
For me, issue consumer is a transient object; it's lifetime is limited to the execution of a single analysis run for a single file, and it should produce a list of all issues for that file.
What should be filtered/suppressed changes independently of the analysis results (e.g. based on suppressions on the server). Filtering/suppression is a view-level operation that should be done every time the view is refreshed i.e. we should have a "model" which contains all the issues for the file which we filter before rendering. It shouldn't be a one-off operation that discards data, since if the filter changes then when we re-apply it we might get the wrong results.